src/coreclr/jit/codegenarmarch.cpp

```diff
@@ -1706,17 +1708,45 @@ void CodeGen::genCodeForIndexAddr(GenTreeIndexAddr* node)
         DWORD scale;
         BitScanForward(&scale, node->gtElemSize);
 
-        // dest = base + index * scale
-        genScaledAdd(emitActualTypeSize(node), node->GetRegNum(), base->GetRegNum(), index->GetRegNum(), scale);
+        if (!index->TypeIs(TYP_I_IMPL))
+        {
+#if defined(TARGET_ARM64)
+            if (scale <= 4)
+            {
+                // target = base + index<<scale
+                GetEmitter()->emitIns_R_R_R_I(INS_add, emitActualTypeSize(node), node->GetRegNum(), base->GetRegNum(),
+                                              indexReg, scale, INS_OPTS_UXTW);
+            }
+            else
+#endif
```

Comment:
The entire `!index->TypeIs(TYP_I_IMPL)` case can be under this #ifdef.

Reply:
(As it is it looks like there is a small arm32 bug, since it will enter the if for a small-typed load)